### PR TITLE
Support Scheduler Go initial service creation

### PR DESCRIPTION
**Pull request by xmlviking**
#57 Support Scheduler in Go

Initial Go version port of the java Scheduler Service.
Features:  Scheduler, SchedulerContext, and Router
Includes:  V2 Configuration changes as per the current EdgeX design pattern
                  Bootstrap, Retry design pattern

TODO: 
- Further enhancement of the Support Scheduler to include additional tests
- Examine the use of the core-metadata as an option for our Event.Addressable needs

Signed-off-by: xmlviking <ecotter@gmail.com>

### CODE REVIEW

**Comment by xmlviking** [comment]
Updated the docker file.  Missed the new file name and the inclusion of the
docker directory now in the projects folder.
I rebuilt it locally and it worked pushed the change up to the PR.

Thanks!

File: cmd/support-scheduler/main.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+	"github.com/edgexfoundry/edgex-go"
+	"github.com/edgexfoundry/edgex-go/internal"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/usage"
+	"github.com/edgexfoundry/edgex-go/internal/support/scheduler"
+	"github.com/edgexfoundry/edgex-go/pkg/clients/logging"
+	"github.com/gorilla/context"
+)
+
+var loggingClient logger.LoggingClient
```
Comment (tsconn23) [inline cmd/support-scheduler/main.go]: Should not be here. Use scheduler.LoggingClient instead.

File: internal/support/scheduler/router.go [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+	"io/ioutil"
+	"net/http"
+)
+
+func LoadRestRoutes() http.Handler {
+	r := mux.NewRouter()
+	b := r.PathPrefix("/api/v1").Subrouter()
+
+	//ping
+	b.HandleFunc("/ping", ping).Methods(http.MethodGet)
```
Comment (tsconn23) [inline internal/support/scheduler/router.go]: Move this above line 22 and use the internal.ApiPingRoute constant. Also, please provide the config endpoint which the SMA will need. Example:
https://github.com/edgexfoundry/edgex-go/blob/master/internal/support/logging/server.go#L211

File: internal/support/scheduler/router.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	b.HandleFunc("/callbacks", removeCallbackAlert).Methods(http.MethodDelete)
+
+	return r
+}
+
+const (
+	ContentTypeKey       = "Content-Type"
+	ContentTypeJsonValue = "application/json; charset=utf-8"
+	ContentLengthKey     = "Content-Length"
+)
```
Comment (tsconn23) [inline internal/support/scheduler/router.go]: These same strings are used pretty widely as literals. Consider moving these constants to internal/constants.go

File: internal/support/scheduler/router.go [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+	if err != nil {
+		LoggingClient.Error("read info request error" + err.Error())
+		return
+	}
+
+	str := `{"name"}: {"` + scheduleEvent.Name + `"}"`
+	str1 := `{"frequency"}:{"` + scheduleEvent.Frequency + `"}"`
+
+	io.WriteString(rw, str+str1)
+	// iterate over the current schedulers
```
Comment (tsconn23) [inline internal/support/scheduler/router.go]: What is this comment for? Is it a TODO?

File: cmd/support-scheduler/main.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	if !ok {
+		logBeforeInit(fmt.Errorf("%s: Service bootstrap failed!", internal.SupportSchedulerServiceKey))
+		return
+	}
+
+	scheduler.LoggingClient.Info("Service dependencies resolved...")
+	scheduler.LoggingClient.Info(fmt.Sprintf("Starting %s %s ", internal.SupportSchedulerServiceKey, edgex.Version))
+
+	// Initialize the  Service Routes
+	scheduler.LoadRestRoutes()
```
Comment (tsconn23) [inline cmd/support-scheduler/main.go]: These routes are being overwritten by the call to startHttpServer on line 67 below. See related comment below.

File: cmd/support-scheduler/main.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+		c := make(chan os.Signal)
+		signal.Notify(c, syscall.SIGINT)
+		errChan <- fmt.Errorf("%s", <-c)
+	}()
+}
+
+func startHttpServer(errChan chan error, port int) {
+	go func() {
+		r := command.LoadRestRoutes()
+		errChan <- http.ListenAndServe(":"+strconv.Itoa(port), context.ClearHandler(r))
```
Comment (tsconn23) [inline cmd/support-scheduler/main.go]: Please note that the routes being served are coming from the command package, not the scheduler package.

File: internal/support/scheduler/router.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+}
+
+func info(rw http.ResponseWriter, req *http.Request) {
+	rw.Header().Set(ContentTypeKey, ContentTypeJsonValue)
+	rw.WriteHeader(http.StatusOK)
+
+	vars := mux.Vars(req)
+
+	scheduleEvent, err := schedulerClient.QueryScheduleWithName(vars["name"])
+	if err != nil {
```
Comment (tsconn23) [inline internal/support/scheduler/router.go]: After fixing the bug which I called out on the main.go file referring to the wrong routes being loaded, I then used Postman to test this method. I issued the following GET `http://localhost:48085/api/v1/info/test` and received a panic. The panic has to do with the schedulerClient not being initialized anywhere in the service and so every function that uses the schedulerClient will panic. Further, what is the intent here? The scheduler service should not be using its client to call itself. Should it be using `pkg/clients/metadata/schedule.go` instead?

```
INFO: 2018/10/06 18:33:39 This is the Support Scheduler Microservice
INFO: 2018/10/06 18:33:39 Service started in: 15.566394ms
INFO: 2018/10/06 18:33:39 Listening on port: 48085
2018/10/06 18:33:45 http: panic serving [::1]:56923: runtime error: invalid memory address or nil pointer dereference
goroutine 6 [running]:
net/http.(*conn).serve.func1(0xc0001cc0a0)
	/usr/local/go/src/net/http/server.go:1746 +0xd0
panic(0x1339080, 0x1636c70)
	/usr/local/go/src/runtime/panic.go:513 +0x1b9
github.com/edgexfoundry/edgex-go/internal/support/scheduler.info(0x1402c00, 0xc0001d6000, 0xc000168900)
	/Users/tconn/go/src/github.com/edgexfoundry/edgex-go/internal/support/scheduler/router.go:57 +0xe2
net/http.HandlerFunc.ServeHTTP(0x13b8768, 0x1402c00, 0xc0001d6000, 0xc000168900)
	/usr/local/go/src/net/http/server.go:1964 +0x44
github.com/edgexfoundry/edgex-go/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc0000ba1e0, 0x1402c00, 0xc0001d6000, 0xc000168900)
	/Users/tconn/go/src/github.com/edgexfoundry/edgex-go/vendor/github.com/gorilla/mux/mux.go:114 +0xe0
github.com/edgexfoundry/edgex-go/vendor/github.com/gorilla/context.ClearHandler.func1(0x1402c00, 0xc0001d6000, 0xc0001d0100)
	/Users/tconn/go/src/github.com/edgexfoundry/edgex-go/vendor/github.com/gorilla/context/context.go:141 +0x8b
net/http.HandlerFunc.ServeHTTP(0xc0000ae920, 0x1402c00, 0xc0001d6000, 0xc0001d0100)
	/usr/local/go/src/net/http/server.go:1964 +0x44
net/http.serverHandler.ServeHTTP(0xc000099450, 0x1402c00, 0xc0001d6000, 0xc0001d0100)
	/usr/local/go/src/net/http/server.go:2741 +0xab
net/http.(*conn).serve(0xc0001cc0a0, 0x1403000, 0xc000020240)
	/usr/local/go/src/net/http/server.go:1847 +0x646
created by net/http.(*Server).Serve
	/usr/local/go/src/net/http/server.go:2851 +0x2f5
```

File: internal/support/scheduler/schedulecontext.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
+//
+// Copyright (c) 2018 Tencent
+//
+// Copyright (c) 2017 Dell Inc
```
Comment (tsconn23) [inline internal/support/scheduler/schedulecontext.go]: Year in copyright is wrong

File: internal/support/scheduler/schedule.go [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+	go func() {
+		for range ticker.C {
+			triggerSchedule()
+		}
+	}()
+}
+
+func StopTicker() {
+	ticker.Stop()
+}
```
Comment (tsconn23) [inline internal/support/scheduler/schedule.go]: This function is unused. Should it be called from init.go::Destruct()?

File: internal/support/scheduler/schedule.go [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+	LoggingClient.Debug("removed the schedule with id : " + scheduleId)
+
+	return nil
+}
+
+//endregion
+
+//region schedule event CRUD
+func queryScheduleEvent(scheduleEventId string) (models.ScheduleEvent, error) {
+	mutex.Lock()
```
Comment (tsconn23) [inline internal/support/scheduler/schedule.go]: GoLand currently identifies this as an unused function. If so, can be removed?

File: internal/support/scheduler/init.go [hunk 11]
```diff
@@ -0,0 +1,265 @@
...
+	go dec.Run()
+
+	select {
+	case <-time.After(2 * time.Second):
+		err = errors.New("timeout loading config from registry")
+	case ex := <-errCh:
+		err = errors.New(ex.Error())
+	case raw := <-updateCh:
+		actual, ok := raw.(*ConfigurationStruct)
+		if !ok {
```
Comment (tsconn23) [inline internal/support/scheduler/init.go]: In order for the scheduler to pull its configuration from Consul, you need to make config-seed aware of the service. I don't see that logic in this PR. Do you intend to include it? If so, you will need to add the respective service key constant to the list found here:
https://github.com/edgexfoundry/edgex-go/blob/master/internal/seed/config/populate.go#L188

File: internal/support/scheduler/init.go [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+				Configuration = actual //Mutex needed?
+			} else {
+				return
+			}
+		}
+	}
+}
+
+func initializeClients(useConsul bool) {
+	// Create metadata clients
```
Comment (tsconn23) [inline internal/support/scheduler/init.go]: This is currently an unused function. Whether it remains that way hinges on the question regarding which service client this application should be using.

File: internal/support/scheduler/init.go [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+func setLoggingTarget() string {
+	if Configuration.Logging.EnableRemote {
+		return Configuration.Clients["Logging"].Url() + clients.ApiLoggingRoute
+	}
+	return Configuration.Logging.File
+}
+
+func AddDefaultSchedulers() error {
+
```
Comment (tsconn23) [inline internal/support/scheduler/init.go]: Consider moving this function to schedule.go. Add unit tests.

File: internal/support/scheduler/init.go [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+	LoggingClient.Info(fmt.Sprintf("loading default schedules and schedule events..."))
+
+	defaultSchedules := Configuration.Schedules
+
```
Comment (tsconn23) [inline internal/support/scheduler/init.go]: Suggest calling `defaultSchedules` just `schedules`

File: internal/support/scheduler/init.go [hunk 15]
```diff
@@ -0,0 +1,265 @@
...
+		}
+		err := addSchedule(defaultSchedule)
+		if err != nil {
+			return LoggingClient.Error("AddDefaultSchedulers() - failed to load schedule %s", err.Error())
+		} else {
+			LoggingClient.Info(fmt.Sprintf("added default schedule %s", defaultSchedule.Name))
+		}
+	}
+
+	defaultScheduleEvents := Configuration.ScheduleEvents
```
Comment (tsconn23) [inline internal/support/scheduler/init.go]: Suggesting calling `defaultScheduleEvents` just `events`

File: internal/support/scheduler/router.go [hunk 16]
```diff
@@ -0,0 +1,231 @@
+//
+// Copyright (c) 2018 Tencent
+//
+// Copyright (c) 2017 Dell Inc.
```
Comment (tsconn23) [inline internal/support/scheduler/router.go]: Year is wrong

File: cmd/support-scheduler/main.go (hunk 1 above)
Comment (xmlviking) [reply cmd/support-scheduler/main.go]: included

File: internal/support/scheduler/router.go (hunk 2 above)
Comment (xmlviking) [reply internal/support/scheduler/router.go]: Updated router to use the mux style incorporated in your logging/server.go example

File: internal/support/scheduler/router.go (hunk 3 above)
Comment (xmlviking) [reply internal/support/scheduler/router.go]: moved local const to scheduler const.go

File: internal/support/scheduler/router.go (hunk 4 above)
Comment (xmlviking) [reply internal/support/scheduler/router.go]: removed comment

File: cmd/support-scheduler/main.go (hunk 6 above)
Comment (xmlviking) [reply cmd/support-scheduler/main.go]: updated the line of code to use the scheduler.LoadRestRoutes()
removed the aforementioned duplication of LoadRestRoutes() as well.

File: internal/support/scheduler/init.go (hunk 11 above)
Comment (xmlviking) [reply internal/support/scheduler/init.go]: I added the service key entry as well as changed the array bounds.

File: internal/support/scheduler/router.go (hunk 7 above)
Comment (xmlviking) [reply internal/support/scheduler/router.go]: Fixed this by removing the client code and adding references to the scheduler functions.
Updated the mux to use the bone style parameters.  
Updated the routes as well to use the style similar to the logging service.

File: internal/support/scheduler/schedule.go (hunk 10 above)
Comment (xmlviking) [reply internal/support/scheduler/schedule.go]: were now using this.

File: internal/support/scheduler/schedule.go (hunk 9 above)
Comment (xmlviking) [reply internal/support/scheduler/schedule.go]: reserved for unit tests which will be added. If there is a better paradigm for adding helper functions for unit tests I would love to do it.

File: internal/support/scheduler/schedulecontext.go (hunk 8 above)
Comment (xmlviking) [reply internal/support/scheduler/schedulecontext.go]: udpated.

File: cmd/support-scheduler/main.go (hunk 5 above)
Comment (xmlviking) [reply cmd/support-scheduler/main.go]: I removed the extra line entry for the scheduler.LoadRestRoutes() and it's now being covered in the startHttpServer

File: internal/support/scheduler/init.go (hunk 12 above)
Comment (xmlviking) [reply internal/support/scheduler/init.go]: removed this function as were not currently using this function.

File: internal/support/scheduler/init.go (hunk 13 above)
Comment (xmlviking) [reply internal/support/scheduler/init.go]: Moved this code into the Scheduler.  Were no longer using a bridge from main to inti too scheduler.

File: internal/support/scheduler/init.go (hunk 14 above)
Comment (xmlviking) [reply internal/support/scheduler/init.go]: removed the references to the default in the schedules and scheduleEvent code.
commit to follow shortly.

File: internal/support/scheduler/init.go (hunk 15 above)
Comment (xmlviking) [reply internal/support/scheduler/init.go]: removed the references to the prefix 'default' from scheduleEvents.  events might be too generic in this case. scheduleEvents is contextual to the use of events.

File: internal/support/scheduler/router.go (hunk 16 above)
Comment (xmlviking) [reply internal/support/scheduler/router.go]: updated the router header information to use the 2018 dell inc.

**Comment by jduranf** [comment]
@xmlviking although you added metadata client in configuration, it's not used. support-scheduler should read all associate schedules from metadata at start up, and add into metadata the schedules defined in the configuration which aren't still in metadata. Also, core-metadata should be modified in order to call support-scheduler `callback` when needed.

**Comment by yanghua** [comment]
Congratulations support-scheduler-go was finally merged. Maybe next time, we keep the initial contributor's git information to make it look better. In addition, edgexfoundry-holding/support-scheduler-go can also be deleted or archived.

**Pull request merged**